src/codegen/compile.ml

```diff
@@ -8931,17 +8982,35 @@ and compile_decs env ae decs captured_in_body : VarEnv.t * scope_wrap =
 and compile_const_exp env pre_ae exp : Const.t * (E.t -> VarEnv.t -> unit) =
   match exp.it with
   | FuncE (name, sort, control, typ_binds, args, res_tys, e) ->
-      let return_tys = match control with
-        | Type.Returns -> res_tys
-        | Type.Replies -> []
-        | Type.Promises -> assert false in
-      let mk_body env ae =
-        List.iter (fun v ->
-          if not (VarEnv.NameEnv.mem v ae.VarEnv.vars)
-          then fatal "internal error: const \"%s\": captures \"%s\", not found in static environment\n" name v
-        ) (Freevars.M.keys (Freevars.exp e));
-        compile_exp_as env ae (StackRep.of_arity (List.length return_tys)) e in
-      FuncDec.closed env sort control name args mk_body return_tys exp.at
+    let fun_rhs =
+
+      (* a few prims cannot be safely inlined *)
+      let inlineable_prim = function
+      | RetPrim -> false
+      | BreakPrim _ -> false
```

Comment:
What about similar ThrowPrim and perhaps even AssertPrim - are there others?

Reply:
Actually, any prim whose implementation/rhs contains such prims should not be naively inlined, unless we, eg,  add a block and an compile RetPrim as Break etc. But perhaps we don't have such beasts.

Reply:
And it looks like ThrowPrim no longer occurs in the final IR - I guess I get rid of it in the async/await translation...

Reply:
> Actually, any prim whose implementation/rhs contains such prims should not be naively inlined

Which is precisely the point of the `inlineable_prim` function…

`ThrowPrim` _should_ be away at this point, but better assert that explicitly…

```suggestion
      | BreakPrim _ -> false
      | ThrowPrim _ -> assert false
```